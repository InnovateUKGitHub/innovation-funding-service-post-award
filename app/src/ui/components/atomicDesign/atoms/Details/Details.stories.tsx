import type { Meta, StoryObj } from "@storybook/react";

import { Info } from "./Details";

const meta: Meta<typeof Info> = {
  title: "GOV.UK Components/Details",
  component: Info,
};

export default meta;

type Story = StoryObj<typeof Info>;

export const Primary: Story = {
  args: {
    summary: "Help with nationality",
    children: `We need to know your nationality so we can work out which elections you’re entitled to vote in. If you cannot provide your nationality, you’ll have to send copies of identity documents through the post.`,
  },
};
