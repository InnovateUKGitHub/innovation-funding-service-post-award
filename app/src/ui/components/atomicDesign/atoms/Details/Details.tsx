import { useGovFrontend } from "@ui/hooks/gov-frontend.hook";

interface InfoProps {
  summary: React.ReactNode;
  children: React.ReactNode;
  qa?: string;
}

export const Info = ({ qa, summary, children }: InfoProps) => {
  const { setRef } = useGovFrontend("Details");

  return (
    <details data-module="govuk-details" className="govuk-details" ref={setRef} data-qa={qa}>
      <summary className="govuk-details__summary">
        <span className="govuk-details__summary-text">{summary}</span>
      </summary>

      <div className="govuk-details__text">{children}</div>
    </details>
  );
};
