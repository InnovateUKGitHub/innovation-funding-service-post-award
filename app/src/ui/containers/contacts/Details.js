import * as React from 'react';
import { connect } from 'react-redux';
import * as Actions from '../../../actions';
import { Link } from 'react-router5';
import Title from '../../layout/Title';
import Breadcrumbs from '../../layout/breadcrumbs';
import * as Tables from '../../tables';
import * as Forms from '../../forms'

class Details extends React.Component {
    componentDidMount() {
        this.props.onLoad(this.props.id);
    }
    render() {
        let data = this.props.data;

        if (!data) {
            return <i>Not Found</i>;
        }

        return (
            <div>
                <Breadcrumbs links={[{ routeName: "home", text: "Home" }, { routeName: "contacts", text: "Contacts" }]}>Contact {data.id}</Breadcrumbs>
                <Title>Contact {data.id}</Title>
                <Forms.Form>
                    <Forms.Fieldset heading="Details">
                        <Forms.ReadOnly label="First Name" name="firstname" value={data.firstName} />
                        <Forms.ReadOnly label="Last Name" name="lastname" value={data.lastName} />
                        <Forms.ReadOnly label="Email" name="email" value={data.email} />
                        <Forms.ReadOnly label="Address" name="address" value={this.renderAddress(data.address)} />
                    </Forms.Fieldset>
                </Forms.Form>
                <Link routeName="contact_edit" routeParams={{ id: data.id }} className="govuk-button govuk-button--start">Edit</Link>
            </div>
        );
    }

    renderAddress(address){
        let items = [address.street, address.city, address.county, address.postcode].filter(x => !!x);
        if(!items.length)
        {
            return "";
        }
        return items.join(", ");
    }
}

function mapStateToProps(state, props) {
    return {
        data: state && state.data && state.data.contact && state.data.contact[props.route.params.id] && state.data.contact[props.route.params.id].data,
        id: props.route.params.id
    };
}

function mapDispachToProps(dispatch) {
    return {
        onLoad: (id) => dispatch(Actions.loadContact(id))
    };
}

export default connect(mapStateToProps, mapDispachToProps)(Details);


