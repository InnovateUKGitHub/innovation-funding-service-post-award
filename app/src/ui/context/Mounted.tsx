import { createContext, useContext, useEffect, useState } from "react";

export interface MountedState {
  isClient: boolean;
  isServer: boolean;
}

export const mountedContext = createContext<MountedState | undefined>(undefined);

/**
 * ###useMountedState
 *
 * hook returns object with isServer and isClient boolean values
 * to show whether app is mounted in the browser (isClient: true)
 */
export function useMountedState(): MountedState {
  const [mounted, setMounted] = useState<boolean>(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  return {
    isServer: !mounted,
    isClient: mounted,
  };
}

interface MountedProviderProps {
  children: React.ReactElement;
}

/**
 * Provider for mounted context
 */
export function MountedProvider(props: MountedProviderProps) {
  const state = useMountedState();

  return <mountedContext.Provider {...props} value={state} />;
}

/**
 * ### useMounted
 *
 * returns the mountedContext state
 */
export function useMounted(): MountedState {
  const state = useContext(mountedContext);

  if (!state) throw Error("You need 'MountedProvider' in order to use this hook.");

  return state;
}

interface IsClientHocProps {
  children: (state: MountedState) => JSX.Element;
}

/**
 * Wrapper react HOC that includes the state of useMounted context
 * and passes it to argument for children
 */
export function MountedHoc({ children }: IsClientHocProps) {
  const state = useMounted();
  return children(state) ?? null;
}
