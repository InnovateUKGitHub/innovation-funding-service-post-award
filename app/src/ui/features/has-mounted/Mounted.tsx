import { createContext, useContext, useEffect, useState } from "react";

import { MountedState } from "./mounted.context";

export const mountedContext = createContext<MountedState | undefined>(undefined);

function useMountedState(): MountedState {
  const [mounted, setMounted] = useState<boolean>(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  return {
    isServer: !mounted,
    isClient: mounted,
  };
}

interface MountedProviderProps {
  children: React.ReactElement;
}

export function MountedProvider(props: MountedProviderProps) {
  const state = useMountedState();

  return <mountedContext.Provider {...props} value={state} />;
}

export function useMounted(): MountedState {
  const state = useContext(mountedContext);

  if (!state) throw Error("You need 'MountedProvider' in order to use this hook.");

  return state;
}

interface IsClientHocProps {
  children: (state: MountedState) => JSX.Element;
}

export function MountedHoc({ children }: IsClientHocProps) {
  const state = useMounted();
  return children(state) ?? null;
}
