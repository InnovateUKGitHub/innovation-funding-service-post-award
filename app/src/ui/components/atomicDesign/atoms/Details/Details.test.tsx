import { render } from "@testing-library/react";
import { MountedProvider } from "@ui/components/atomicDesign/atoms/providers/Mounted/Mounted";

import { Info } from "./Details";

describe("<Info />", () => {
  describe("@renders", () => {
    test("with string contents", () => {
      const { container } = render(
        <MountedProvider>
          <Info summary="Slimming World">Did you know that apples are mostly red?</Info>
        </MountedProvider>,
      );

      expect(container).toMatchSnapshot();
    });

    test("with ReactNode contents", () => {
      const { container } = render(
        <MountedProvider>
          <Info
            summary={
              <>
                Slimming <b>world</b>
              </>
            }
          >
            <h1>Sus Individuals</h1>
            <ul>
              <li>Red</li>
              <li>Orange</li>
            </ul>
          </Info>
        </MountedProvider>,
      );

      expect(container).toMatchSnapshot();
    });
  });
});
