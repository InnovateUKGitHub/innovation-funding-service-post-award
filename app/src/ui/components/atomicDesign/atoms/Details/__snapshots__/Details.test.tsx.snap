// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Info /> @renders with ReactNode contents 1`] = `
<div>
  <details
    class="govuk-details"
    data-module="govuk-details"
  >
    <summary
      class="govuk-details__summary"
    >
      <span
        class="govuk-details__summary-text"
      >
        Slimming 
        <b>
          world
        </b>
      </span>
    </summary>
    <div
      class="govuk-details__text"
    >
      <h1>
        Sus Individuals
      </h1>
      <ul>
        <li>
          Red
        </li>
        <li>
          Orange
        </li>
      </ul>
    </div>
  </details>
</div>
`;

exports[`<Info /> @renders with string contents 1`] = `
<div>
  <details
    class="govuk-details"
    data-module="govuk-details"
  >
    <summary
      class="govuk-details__summary"
    >
      <span
        class="govuk-details__summary-text"
      >
        Slimming World
      </span>
    </summary>
    <div
      class="govuk-details__text"
    >
      Did you know that apples are mostly red?
    </div>
  </details>
</div>
`;
